### DISCUSSION TITLE
String array value not encoded correctly?

### DISCUSSION THREAD

**Discussion by doolio** [joaotavora/eglot#1048] [Q&A] [answered]
The[ `python-lsp-server` ](https://github.com/python-lsp/python-lsp-server/blob/develop/CONFIGURATION.md)has several configuration settings whose dataype is an array of strings. For some the strings must be unique for others they do not need to be. For all but one of these settings the default value is either `null` (i.e. no value is set instead the server will use the plugins default value) or `[]` (i.e. the empty array). The `pylsp.plugins.jedi_completion.cache_for` setting has a non-empty array of strings as its default value namely `["pandas", "numpy", "tensorflow", "matplotlib"]`.

In a `.dir-local.el` file I have this defined as follows. Note my `.dir-locals.el` file contains all other available server settings but I have omitted those below. I include some inline comments as well.

```
((python-mode
  . ((eglot-workspace-configuration
      . ((:pylsp
          . (:plugins
             (:jedi_completion
              (:cache_for ["pandas", "numpy", "tensorflow", "matplotlib"] ; string array: ["pandas", "numpy", "tensorflow", "matplotlib"] (default)
                          :eager :json-false ; boolean: true or false (default)
                          :enabled t ; boolean: true (default) or false
                          :fuzzy :json-false ; boolean: true or false (default)
                          :include_class_objects t ; boolean: true (default) or false
                          :include_function_objects t ; boolean: true (default) or false
                          :include_params t ; boolean: true (default) or false
                          :resolve_at_most 25))))))))) ; integer: 25 (default)
``` 

However, checking either the `eglot-workspace-configuration` variable or via the newly added `M-x eglot-show-workspace-configuration` command this seems to be encoded as follows:

```
{
  "pylsp": {
    "plugins": {
      "jedi_completion": {
        "cache_for": [
          "pandas",
          {
            ",": "numpy"
          },
          {
            ",": "tensorflow"
          },
          {
            ",": "matplotlib"
          }
        ],
        "eager": false,
        "enabled": true,
        "fuzzy": false,
        "include_class_objects": true,
        "include_function_objects": true,
        "include_params": true,
        "resolve_at_most": 25
      }
    }
  }
}
```
This seems incorrect to me. It seems there is an extra `,` (4 vs 3) and after the first one the other`,`'s are being treated as keys. I have this setting as the first entry in a (nested) plist for the `pylsp.plugins.jedi_completion` settings so they are listed in alphabetical order. I don't think the order matters within a plist. In any case, with this setting first in the plist as shown above and as I believe encoded incorrectly it means the remaining settings don't get acted upon by the server. I believe they are sent to the server but just not acted upon. For me this means I get no completion candidates despite the `"enabled": true` above.

Is this a misconfiguration on my part or an edge case the recent discussions around PRs #790 and #1033 have not considered?

I've also noticed (but this may be a separate issue is that if I change the settings in my `.dir-locals.el` file and then `M-x eglot-signal-didChangeConfiguration` my changes don't get sent to the server or at least no response is received from the server. The signal it seems is sent but it just resends the settings given by `eglot-workspace-configuration` variable which has not updated with the changes I made. Would it be worthwhile if the `eglot-signal-didChangeConfiguration` provided some feedback in the echo area?

**Comment by fbergroth** [comment, marked as answer]
There should not be `,` between elements in the vector, try:
```elisp
(:cache_for ["pandas" "numpy" "tensorflow" "matplotlib"] 
```

I usually reload .dir-locals by running revert-buffer, but perhaps there are better ways. What feedback would have helped you in this case?

**Reply by doolio** [reply]
Thank you removing the `,` in the vector did the trick.

I was simply saving the `dir-locals.el` file after making a change and executing `M-x eglot-signal-didChangeConfiguration`. I ran `revert-buffer` as you suggested and then sent the signal but to no avail - my changes were not sent to the server. I had to shutdown the server and restart it to pick them up.

Edit: In terms of feedback an indication that the signal was not acknowledged by the server perhaps.

**Reply by fbergroth** [reply]
> I ran `revert-buffer` as you suggested and then sent the signal but to no avail - my changes were not sent to the server.

Strange. If I run `C-u M-x eglot` and type in `pylsp -v`, then I can observe `pylsp.config.config - Updated settings to ...` whenever i run `M-x eglot-signal-didChangeConfiguration`. Perhaps the settings you changed are only read during initialization?

**Reply by doolio** [reply]
Even enabling a more verbose output where I too can see the `pylsp.config.config - Updated settings to ...` type messages a change I make to a setting is not updated. It still retains the old value. If anything it appears the `pylsp.config.config - Updated settings to ...` type messages are repeated with what appears to be the same information. Not sure why that happens.

I don't think these are settings read only during initialization but how can you tell?
